internal/server/auth/method/oidc/server_test.go

```diff
@@ -0,0 +1,299 @@
+		err = protojson.Unmarshal(body, &authorize)
+		require.NoError(t, err)
+
+		authURL, err = url.Parse(authorize.AuthorizeUrl)
+		require.NoError(t, err)
+	})
+
+	t.Log("Navigating to authorize URL:", authURL.String())
+
+	var location string
+	t.Run("Login as Mark", func(t *testing.T) {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, authURL.String(), nil)
+		require.NoError(t, err)
+
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+
+		values, err := parseLoginFormHiddenValues(resp.Body)
+		require.NoError(t, err)
+
+		// add login credentials
+		values.Set("uname", "mark")
+		values.Set("psw", "phelps")
+
+		req, err = http.NewRequestWithContext(ctx, http.MethodPost, tp.Addr()+"/login", strings.NewReader(values.Encode()))
+		require.NoError(t, err)
+
+		req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+		resp, err = client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+
+		require.Equal(t, http.StatusFound, resp.StatusCode)
+
+		location = resp.Header.Get("Location")
+	})
+
+	t.Log("Redirecting to callback URL:", location)
+
+	t.Run("Callback", func(t *testing.T) {
+		req, err := http.NewRequestWithContext(ctx, http.MethodGet, location, nil)
+		require.NoError(t, err)
+
+		resp, err := client.Do(req)
+		require.NoError(t, err)
+		defer resp.Body.Close()
+
+		data, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		var response auth.CallbackResponse
+		if !assert.NoError(t, protojson.Unmarshal(data, &response)) {
+			t.Log("Unexpected response", string(data))
+			t.FailNow()
+		}
+
+		assert.Empty(t, response.ClientToken) // middleware moves it to cookie
+		assert.Equal(t, auth.Method_METHOD_OIDC, response.Authentication.Method)
+		assert.Equal(t, map[string]string{
+			"io.flipt.auth.id.email":      "mark@flipt.io",
+			"io.flipt.auth.oidc.provider": "OIDC_PROVIDER_GOOGLE",
+		}, response.Authentication.Metadata)
+
+		// ensure expiry is set
+		assert.NotNil(t, response.Authentication.ExpiresAt)
+
+		// obtain returned cookie
+		cookie, err := (&http.Request{
+			Header: http.Header{"Cookie": resp.Header["Set-Cookie"]},
+		}).Cookie("flipt_client_token")
+		require.NoError(t, err)
+
+		// check authentication in store matches
+		storedAuth, err := server.GRPCServer.Store.GetAuthenticationByClientToken(ctx, cookie.Value)
+		require.NoError(t, err)
+
+		// ensure stored auth can be retrieved by cookie abd matches response body auth
+		if diff := cmp.Diff(storedAuth, response.Authentication, protocmp.Transform()); err != nil {
+			t.Errorf("-exp/+got:\n%s", diff)
+		}
+	})
+}
+
+// parseLoginFormHiddenValues parses the contents of the supplied reader as HTML.
+// It descends into the document looking for the hidden values associated
+// with the login form.
+// It collecs the hidden values into a url.Values so that we can post the form
+// using our Go client.
+func parseLoginFormHiddenValues(r io.Reader) (url.Values, error) {
```

Comment:
😭 It definitely either sits-on or leaps over the line of acceptability 😂 

I wanted to avoid bringing in more dependencies if I could. I just need to pull a particular form's hidden inputs out of an HTML page. But perhaps using a library would make this a lot more digestible.
What I really wanted was that test library to offer a way to get the `code` parameter without parsing HTML. I just haven't found it yet.
I guess, at-least, this is a little closer to what a user in a browser would do. Slightly more representative.